app/controllers/tasks_controller.rb

```diff
@@ -0,0 +1,82 @@
+class TasksController < ApplicationController
+  
+  def index 
+    @tasks = Task.all
+  end
+  
+  def show 
+    task_id = params[:id]
+    @task = Task.find_by(id: task_id)
+    if @task.nil?
+      redirect_to tasks_path
+      return 
+    end 
+  end
+  
+  def new
+    @task = Task.new
+  end
+  
+  def create
+    @task = Task.new(name: params[:task][:name], description: params[:task][:description]) 
+    if @task.save 
+      redirect_to task_path(@task.id)
+      return
+    else 
+      render :new 
+      return
+    end
+  end
+  
+  def edit
+    @task = Task.find_by(id: params[:id])
+    if @task.nil?
+      redirect_to tasks_path
+      return
+    end
+  end
+  
+  def update
+    @task = Task.find_by(id: params[:id])
+    
+    if @task.nil?
+      redirect_to tasks_path
+      return 
+    elsif @task.update(name: params[:task][:name], description: params[:task][:description], completion_date: params[:task][:completion_date])
+      redirect_to tasks_path
+      return 
+    else 
+      render :edit
+      return
+    end
+  end 
+  
+  def completed 
```

Comment:
Consider how making two separate methods, for instance `mark_complete` and `mark_incomplete` would make this action idempotent.